### PR TITLE
fix(db): cannot replace original pl name where editing pl

### LINKED ISSUE #3941: [Bug]: Error reported when modifying a function with special characters as names
### ODC version

odc433

### OB version

*

### What happened?

image

### What did you expect to happen?

Modified successfully

### How can we reproduce it (as minimally and precisely as possible)?

1. Create Function
create function `FuN_tesT@111~!#$%^&*()_+` (`p1` int(11)) returns int(11) begin declare v1 int; set v1 = p1 + 1; return v1; end;
2. Modify function

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by zijiacj**
#### What type of PR is this?
type-bug

#### What this PR does / why we need it:
For the name `FuN_tesT@111~! The #$%^&*()_+` function, which contains special regular expression characters, was recognized as a regular match using `String.replaceFirst` method, resulting in an uncompleted replacement. 

Invalid the special characters of the regular expression contained in the function name, then use the `replaceFirst` method to complete the replacement
#### Which issue(s) this PR fixes:

Fixes #3941 

#### Additional documentation e.g., usage docs, etc.:

```docs

```

### CODE REVIEW

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java [hunk 1]
```diff
@@ -99,7 +99,7 @@ private EditPLResp executeWrappedEditPL(String sessionId, EditPLReq editPLReq,
             DBObjectType plType, String tempPlName) throws Exception {
         String plName = editPLReq.getObjectName();
         String editPLSql = editPLReq.getSql();
-        String tempPLSql = editPLSql.replaceFirst(plName, tempPlName);
+        String tempPLSql = editPLSql.replaceFirst(plName.replaceAll("([\\\\+*?\\[\\](){}|.^$])", "\\\\$1"), tempPlName);
```
Comment (PeachThinking) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java:102]: It is recommended to write the regular expression as a separate attribute

Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java:102]: how about this？
image
image

Comment (PeachThinking) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java:102]: 1. `REGULAR_CHARS` can be `String[]`?
2. It is better not to return `null`, you can return `Optional`

Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java:102]: private static final char[] REGULAR_CHARS = { '.', '*', '+', '?', '^', '$', '[', ']', '(', ')', '{', '}', '|', '\\' };

Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java:102]: private static final char[] REGULAR_CHARS = { '.', '*', '+', '?', '^', '$', '[', ']', '(', ')', '{', '}', '|', '\\' };

Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java:102]: 2. return ""
image

File: server/odc-common/src/test/java/com/oceanbase/odc/common/util/StringUtilsTest.java [hunk 2]
```diff
@@ -398,4 +400,31 @@ public void startWithIgnoreSpaceAndNewLines_targetsNonStartingWithPrefixWithMaxM
...
+    public void replaceFirst_plEditReplaceOriginalName_success() {
+        String pl =
+                "create function `FuN_test@111--#%&*()_+` (`p1` int(12)) returns int(11) begin declare v1 int; set v1 = p1 + 1; return v1; end;";
+        String originalName = "FuN_test@111--#%&*()_+";
+        String replaceName = "replace_name";
+        String expect =
+                "create function `replace_name` (`p1` int(12)) returns int(11) begin declare v1 int; set v1 = p1 + 1; return v1; end;";
+        Optional<String> escapedRegex = StringUtils.escapeRegex(originalName);
+        String actual = pl.replaceFirst(escapedRegex.get(), replaceName);
+        Assert.assertEquals(expect, actual);
```
Comment (PeachThinking) [inline server/odc-common/src/test/java/com/oceanbase/odc/common/util/StringUtilsTest.java:427]: why the expect function name is "replace_name"?
the expect function name should be "Optional[FuN_test@111--#%&\*\(\)_\+]"？

Comment (zijiacj) [reply server/odc-common/src/test/java/com/oceanbase/odc/common/util/StringUtilsTest.java:427]: That's no problem.That is mainly to test the `repalceFirst` method. The first occurrence of the `originalName` in the `pl` below needs to be replaced
image

Comment (zijiacj) [reply server/odc-common/src/test/java/com/oceanbase/odc/common/util/StringUtilsTest.java:427]: The test method name has changed to` replaceFirst_replaceOriginalNameContainingRegularCharsInPlEdit_success`

**Review by PeachThinking — APPROVED** [review]
lgtm

**Review by guowl3 — APPROVED** [review]
lgtm

**Pull request merged**